R/methods-CollectionWithMetadata.R

```diff
@@ -0,0 +1,247 @@
+#' @return vector of all ID columns
+#' @rdname getIdColumns
+#' @export
+#' @include class-CollectionWithMetadata.R
+setGeneric("getIdColumns",
+    function(object) standardGeneric("getIdColumns"),
+    signature = c("object")
+)
+
+#' @rdname getIdColumns
+#' @aliases getIdColumns,ANY-method
+setMethod("getIdColumns", "ANY", function(object) {
+    if (all(c('recordIdColumn','ancestorIdColumns') %in% slotNames(object))) {
+        return(c(object@recordIdColumn, object@ancestorIdColumns))
+    } else {
+        stop("Object does not have recordIdColumn and/or ancestorIdColumns slots. Received object of class ", class(object))
+    }
+})
+
+#' Get Variable Names of Metadata
+#' 
+#' Get the names of the metadata variables in an object containing sample metadata.
+#' @param object Any object with a slot containing sampleMetadata
+#' @return a character vector of metadata variable names
+#' @rdname getMetadataVariableNames
+#' @export
+setGeneric("getMetadataVariableNames", function(object) standardGeneric("getMetadataVariableNames"))
+
+#' @rdname getMetadataVariableNames
+#' @aliases getMetadataVariableNames,CollectionWithMetadata-method
+setMethod("getMetadataVariableNames", "CollectionWithMetadata", function(object) return(names(object@sampleMetadata@data)))
+
+#' Get Sample Metadata Id Column Names
+#' 
+#' Get the names of the record and ancestor id columns in the sample metadata of an object.
+#' @param object An object w sample metadata
+#' @return a character vector of id column names
+#' @rdname getSampleMetadataIdColumns
+#' @export
+setGeneric("getSampleMetadataIdColumns", function(object) standardGeneric("getSampleMetadataIdColumns"))
+
+#' @rdname getSampleMetadataIdColumns
+#' @aliases getSampleMetadataIdColumns,CollectionWithMetadata-method
+setMethod("getSampleMetadataIdColumns", "CollectionWithMetadata", function(object) getIdColumns(object@sampleMetadata))
+
+#' Get data.table of values from CollectionWithMetadata
+#'
+#' Returns a data.table of collection values, respecting the
+#' `imputeZero` slot.
+#' 
+#' @param object CollectionWithMetadata
+#' @param ignoreImputeZero boolean indicating whether we should respect the imputeZero slot
+#' @param includeIds boolean indicating whether we should include recordIdColumn and ancestorIdColumns
+#' @param verbose boolean indicating if timed logging is desired
+#' @return data.table of values
+#' @rdname getCollectionValues
+#' @export
+setGeneric("getCollectionValues",
+    function(object, ignoreImputeZero = c(FALSE, TRUE), includeIds = c(TRUE, FALSE), verbose = c(TRUE, FALSE)) standardGeneric("getCollectionValues"),
+    signature = c("object")
+)
+
+#' @rdname getCollectionValues
+#' @aliases getCollectionValues,CollectionWithMetadata-method
+setMethod("getCollectionValues", signature("CollectionWithMetadata"), function(object, ignoreImputeZero = c(FALSE, TRUE), includeIds = c(TRUE, FALSE), verbose = c(TRUE, FALSE)) {
```

Comment:
how about getCollectionData? getCollectionValues i expected to return a unique list of values in the collection